cmd/controller/main.go

```diff
@@ -19,11 +19,15 @@ import (
 
 	"github.com/aws/karpenter/pkg/cloudprovider"
 	"github.com/aws/karpenter/pkg/cloudprovider/aws"
+	awscontrollers "github.com/aws/karpenter/pkg/cloudprovider/aws/controllers"
 	"github.com/aws/karpenter/pkg/controllers"
 )
 
 func main() {
-	controllers.Initialize(func(ctx context.Context, options cloudprovider.Options) cloudprovider.CloudProvider {
-		return aws.NewCloudProvider(ctx, options)
+	controllers.Initialize(func(ctx context.Context, options cloudprovider.Options) (cloudprovider.CloudProvider, controllers.ControllerInitFunc) {
+		provider := aws.NewCloudProvider(ctx, options)
+		return provider, func(c context.Context, opts *controllers.ControllerOptions) []<-chan struct{} {
```

Comment:
Can you clarify the part that doesn't seem straightforward to you? Is it the cleanup portion or just the injection in general?

Reply:
Can you clarify the part that doesn't seem straightforward to you? Is it the cleanup portion or just the injection in general?

Reply:
The interface we are providing to a cloud provider is that they give us:
1) A function
2) That returns a function
3) That returns a channel of structs.

To figure out how to use that you're going to have to spend a good amount of time.

We should probably be using the controller-manager's `Runnable` and `LeaderElectionRunnable` interfaces both internally and for allowing cloudproviders to register their own application lifetime related processing.  That would hook them up into the lifetime that the controller-manager is already managing.

Reply:
I went back and forth between exposing Runnable on the CloudProvider interface so it was more discoverable; however, there was a dependency issue with injecting things like ClusterState into these controllers since cluster state uses the CloudProvider interface. There may be a better way to decouple these things but, in my mind, it's not unreasonable to think that cloudproviders might rely on cluster-state, and potentially other components from the common controller set/packages in the future.

The interface is simplified now to they give us:
1. A function
2. That returns a function that initializes the controllers

Fundamentally though, at this point, we need the concept of a `CloudProvider` interface and a `CloudProviderControllerSet` interface and they have to be completely decoupled for the dependency tree to work